File: src/org/cerion/symcalc/expression/function/N.java
package org.cerion.symcalc.expression.function;

import org.cerion.symcalc.Environment;
import org.cerion.symcalc.expression.Expr;
import org.cerion.symcalc.expression.FunctionExpr;
import org.cerion.symcalc.expression.number.IntegerNum;

public class N extends FunctionExpr {

    public N(Expr... e) {
        super(FunctionType.N,e);
    }

    @Override
    protected Expr evaluate() {
        if(size() > 0) {
            if(size() > 1 && get(1).isInteger()) {
                IntegerNum n = (IntegerNum)get(1);
                getEnv().setNumericalEval(true, n.toInteger());
            } else
                getEnv().setNumericalEval(true);

            return get(0).eval();
        }

        return this;
    }
}

File: src/expression/function/core/N.kt
package org.cerion.symcalc.expression.function.core

import org.cerion.symcalc.expression.ConstExpr
import org.cerion.symcalc.expression.Expr
import org.cerion.symcalc.expression.function.FunctionExpr
import org.cerion.symcalc.expression.function.FunctionFactory
import org.cerion.symcalc.expression.number.NumberExpr
import org.cerion.symcalc.expression.number.RealBigDec

class N(vararg e: Expr) : FunctionExpr(*e) {

    override fun evaluate(): Expr {
        val e = args[0]
        val machinePrecision = size == 1
        val precision = if (machinePrecision) MachinePrecision else get(1).asInteger().intValue()

        if (precision == 0)
            return RealBigDec.ZERO

        if (e is ConstExpr)
            return e.evaluate(precision)
        if (e is NumberExpr)
            return e.evaluate(precision)

        val newArgs = mutableListOf<Expr>()
        for (i in 0 until e.args.size) {
            newArgs.add(e.args[i].eval(precision))
        }

        if(e is FunctionExpr) {
            return FunctionFactory.createInstance(e.name, *newArgs.toTypedArray()).eval()
        }

        return this
    }

    override fun validate() {
        validateParameterRage(1, 2)
    }
}
